HW3/P3/sum.cl

```diff
@@ -24,8 +28,15 @@ __kernel void sum_coalesced(__global float* x,
     // You can assume get_local_size(0) is a power of 2.
     //
     // See http://www.nehalemlabs.net/prototype/blog/2014/06/16/parallel-programming-with-opencl-and-python-parallel-reduce/
-    for (;;) { // YOUR CODE HERE
-        ; // YOUR CODE HERE
+    j=1;
+    offset=get_local_size(0)>>j;
+    while(offset>1) {
+    //for (j = 1; j <= lim; j++) { // YOUR CODE HERE
+        offset=get_local_size(0)>>j;
+        //printf("global_id[%d], offset=%d\n",global_id,offset);
+        fast[local_id]=fast[local_id]+fast[local_id+offset];   
```

Comment:
If you compare this to the code in the link above, you'll see that an if statement is missing.
This causes 2 issues:
1. Some threads may access data that is beyond the end of the input array.
2. The output may differ because in some iteration, given two threads A and B (where B_id = A_id + offset), B may update its value, and then A will read the new value.
This may also explain why you saw an error when using more than 32 threads, or when the global size was larger than 4096.